### PR TITLE
Try to add cachix support

### PR DESCRIPTION
**Pull request by Gerschtli**
Hey, I just open a new PR for cachix support as we started to discuss in https://github.com/t184256/nix-on-droid-bootstrap/pull/13.

I tested it on my computer with following steps:

1. Build and push store paths with `nix-build ci.nix | cachix push gerschtli`
2. Remove generated gcroots:
```bash
rm result*
nix-collect-garbage
```
3. Build and fetch store paths again with `nix-build ci.nix` (cachix binary cache has to be registered locally already):
```
copying path '/nix/store/vl48fgr1v9j2lsncxlwh3ifzh2h9w70r-source' from 'https://cache.nixos.org'...
building '/nix/store/5bibwf4y9lz8c952q7q18rv4nyh35jhx-nix-installer-sha256.drv'...
unpacking sources
patching sources
configuring
no configure script, doing nothing
building
no Makefile, doing nothing
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256
strip is /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/strip
patching script interpreter paths in /nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256
checking for references to /build/ in /nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256...
copying path '/nix/store/3izx77j932vi5g03725rv1qmcdjmsvjn-nixDirectory' from 'https://gerschtli.cachix.org'...
building '/nix/store/l00q0xkj8m3mk4vp0k6blszqhjhdxbaz-nix-installer-sha256.drv'...
unpacking sources
patching sources
configuring
no configure script, doing nothing
building
no Makefile, doing nothing
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256
strip is /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/strip
patching script interpreter paths in /nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256
checking for references to /build/ in /nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256...
copying path '/nix/store/zyxl5gmrhy9gql4cw7hpv05hz947aal0-nixDirectory' from 'https://gerschtli.cachix.org'...
these derivations will be built:
  /nix/store/23bwyp0gb2r21abp93nzhznllycg2vs5-login-inner.drv
  /nix/store/5i6qpb3ypr10wlg6kvi56w5syifkwcwb-nix.conf.drv
  /nix/store/7pyzz8b723q3m78606c9xap5c614gihr-nix-on-droid-install.drv
  /nix/store/cz538illxw8cr4a2khvq8gnv2yab9z0w-login.drv
  /nix/store/fflhixzb56iyi94rv0jcaqchsqq6ajbs-nix-on-droid-install.drv
  /nix/store/kby963p6gbm56w43zj4i7yxfwfznzhmb-login-inner.drv
  /nix/store/rp8yadwvsjxpan0lc11h5pc7kxsh1dh9-home.nix.default.drv
  /nix/store/wgch6p4mf35an2g1djnqa8gz9253dlsm-resolv.conf.drv
these paths will be fetched (394.33 MiB download, 2153.35 MiB unpacked):
  /nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap
  /nix/store/0xy2280dk772d64kvba3ibvcapqn5vl5-platform-tools-28.0.1
  /nix/store/1930ls2j0ay6pgk1lzf5fnj6qsfrkffx-stdenv-linux
  /nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap
  /nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android
  /nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android
  /nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android
  /nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip
  /nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android
  /nix/store/hx9mk75m6xb4vc1mczd48pdcmp5if95k-bionic-prebuilt
  /nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip
  /nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static
  /nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045
  /nix/store/qfg481dl6wal1ik4z7xqlgjnhjkxrkr8-bionic-prebuilt
  /nix/store/rryr4ihjnnwxr6qlj5x9p1mwyg82d34l-libc++abi-5.0.2
  /nix/store/s2f809sapq9dcbbgl1gv2yfra119s740-libc++-5.0.2
  /nix/store/wnxw0bi0jcr729akrh8sqpizdvhdnh9g-ncurses-6.1-20190112-abi5-compat
copying path '/nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/rryr4ihjnnwxr6qlj5x9p1mwyg82d34l-libc++abi-5.0.2' from 'https://cache.nixos.org'...
copying path '/nix/store/wnxw0bi0jcr729akrh8sqpizdvhdnh9g-ncurses-6.1-20190112-abi5-compat' from 'https://cache.nixos.org'...
copying path '/nix/store/s2f809sapq9dcbbgl1gv2yfra119s740-libc++-5.0.2' from 'https://cache.nixos.org'...
copying path '/nix/store/0xy2280dk772d64kvba3ibvcapqn5vl5-platform-tools-28.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/1930ls2j0ay6pgk1lzf5fnj6qsfrkffx-stdenv-linux' from 'https://cache.nixos.org'...
building '/nix/store/rp8yadwvsjxpan0lc11h5pc7kxsh1dh9-home.nix.default.drv'...
building '/nix/store/23bwyp0gb2r21abp93nzhznllycg2vs5-login-inner.drv'...
building '/nix/store/kby963p6gbm56w43zj4i7yxfwfznzhmb-login-inner.drv'...
building '/nix/store/cz538illxw8cr4a2khvq8gnv2yab9z0w-login.drv'...
building '/nix/store/7pyzz8b723q3m78606c9xap5c614gihr-nix-on-droid-install.drv'...
building '/nix/store/fflhixzb56iyi94rv0jcaqchsqq6ajbs-nix-on-droid-install.drv'...
building '/nix/store/5i6qpb3ypr10wlg6kvi56w5syifkwcwb-nix.conf.drv'...
building '/nix/store/wgch6p4mf35an2g1djnqa8gz9253dlsm-resolv.conf.drv'...
copying path '/nix/store/hx9mk75m6xb4vc1mczd48pdcmp5if95k-bionic-prebuilt' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/qfg481dl6wal1ik4z7xqlgjnhjkxrkr8-bionic-prebuilt' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android' from 'https://gerschtli.cachix.org'...
/nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap
/nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip
/nix/store/n0cwr7553h1qscs7fg3yi5ak0s0kv911-home.nix.default
/nix/store/13f9dn1dcz6v3nr0rkbg5mi7m0wr23cx-login
/nix/store/nkm4zafhqgggvmdk0fvn2y5vx1p508rd-login-inner
/nix/store/bj6wbzdf6f4aww23l1dbxalsdsgjqm61-nix.conf
/nix/store/01xidc3kn1q2dgr1hs61wbjp8lkpjg45-nix-on-droid-install
/nix/store/68c85jg2gf0kkwlxhvd6rd5rxph9d9z6-resolv.conf
/nix/store/3izx77j932vi5g03725rv1qmcdjmsvjn-nixDirectory
/nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256
/nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android
/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static
/nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android
/nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap
/nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip
/nix/store/n0cwr7553h1qscs7fg3yi5ak0s0kv911-home.nix.default
/nix/store/13f9dn1dcz6v3nr0rkbg5mi7m0wr23cx-login
/nix/store/ym08s9gw48kx3nrsl6rxjvhrj0jsfvvm-login-inner
/nix/store/bj6wbzdf6f4aww23l1dbxalsdsgjqm61-nix.conf
/nix/store/lvi50fymscyj4m1a8h12ia9j6m8xxbak-nix-on-droid-install
/nix/store/68c85jg2gf0kkwlxhvd6rd5rxph9d9z6-resolv.conf
/nix/store/zyxl5gmrhy9gql4cw7hpv05hz947aal0-nixDirectory
/nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256
/nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android
/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static
/nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android
```

The fact that some derivations still get built locally is caused by two factors:
- the impurity of the `nix-installer-sha256` because it downloads the sha256 file everytime it gets built
- the derivations built via `writeText` and `writeScript`. I was not able to figure out yet why these get pushed but don't get fetched from binary cache.. As you can see the depending derivations like `bootstrap` are fetched, which means the store path is identically to the ones I already pushed. I assume that some logic in nixpkgs decides that simply rebuilding a `writeText`/`writeScript` is more efficient than downloading it. I'm going to dig further!

I tried `nix-build ci.nix` on another machine and got the following output:
```
copying path '/nix/store/vl48fgr1v9j2lsncxlwh3ifzh2h9w70r-source' from 'https://cache.nixos.org'...
copying path '/nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/3izx77j932vi5g03725rv1qmcdjmsvjn-nixDirectory' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/zyxl5gmrhy9gql4cw7hpv05hz947aal0-nixDirectory' from 'https://gerschtli.cachix.org'...
these derivations will be built:                                                                         
  /nix/store/23bwyp0gb2r21abp93nzhznllycg2vs5-login-inner.drv                            
  /nix/store/5i6qpb3ypr10wlg6kvi56w5syifkwcwb-nix.conf.drv                            
  /nix/store/7pyzz8b723q3m78606c9xap5c614gihr-nix-on-droid-install.drv
  /nix/store/cz538illxw8cr4a2khvq8gnv2yab9z0w-login.drv    
  /nix/store/fflhixzb56iyi94rv0jcaqchsqq6ajbs-nix-on-droid-install.drv
  /nix/store/kby963p6gbm56w43zj4i7yxfwfznzhmb-login-inner.drv                                                                                                                                                                                                                                                                                     
  /nix/store/rp8yadwvsjxpan0lc11h5pc7kxsh1dh9-home.nix.default.drv
  /nix/store/wgch6p4mf35an2g1djnqa8gz9253dlsm-resolv.conf.drv                                                                                                                                                                                                                                                                                     
these paths will be fetched (476.73 MiB download, 2360.59 MiB unpacked):
  /nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap                                                                                                                                                                                                                                                                                           
  /nix/store/0xy2280dk772d64kvba3ibvcapqn5vl5-platform-tools-28.0.1                                                                                                                                                                                                                                                                               
  /nix/store/2iqzwckaprdrr69wy9kap23i94q0bi3w-openjdk-8u212-ga   
  /nix/store/2p5331cg724c69hcfj203ji8sj9bqv2s-polkit-0.115        
  /nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap              
  /nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android                                                                                                                                                                                                                                         
  /nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android
  /nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android                                                                                                                                                                                                                                                            
  /nix/store/7x5sfjddnq51py9h7hdj3369dwnlnfkh-ORBit2-2.14.19                                                                                                                                                                                                                                                                                      
  /nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip                                                                                                                                                                                                                                                                                       
  /nix/store/bqyn2l33zzm8pgcd973g46qz534swzm3-liberation-fonts-2.00.4                                                                                                                                                                                                                                                                             
  /nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android
  /nix/store/dz580wi3mhj0a217vlyj1hdih3ldiz14-dbus-glib-0.110                                              
  /nix/store/g3qsai0nz8maad51349v19yqf16rj2yv-alsa-lib-1.1.8                                               
  /nix/store/hx9mk75m6xb4vc1mczd48pdcmp5if95k-bionic-prebuilt                                              
  /nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip                                                    
  /nix/store/jsycgv6la5nhsyb2034l7xfc5sa0ajqx-gconf-3.2.6                                                      
  /nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static                                           
  /nix/store/nqf6dmfdb7b62bdapysnf2dc4cjp7al1-openjdk-8u212-ga-jre                               
  /nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045                                     
  /nix/store/qfg481dl6wal1ik4z7xqlgjnhjkxrkr8-bionic-prebuilt                                               
  /nix/store/rg93lv1yxxqw3v9hg1hkpfqr2nab5p7k-hook                                                                                                                                                                                                                                                                                                
  /nix/store/rryr4ihjnnwxr6qlj5x9p1mwyg82d34l-libc++abi-5.0.2                                                                                                                                                                                                                                                                                     
  /nix/store/s2f809sapq9dcbbgl1gv2yfra119s740-libc++-5.0.2                                                          
  /nix/store/vzmvbd94qm3kjv8m7ljbsyy8n3nygmy6-gnome-vfs-2.24.4                                                               
  /nix/store/wnxw0bi0jcr729akrh8sqpizdvhdnh9g-ncurses-6.1-20190112-abi5-compat                                    
  /nix/store/x8ly9dpv355jndl8wn3br2l50n3kz1s2-libIDL-0.8.14 
copying path '/nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/g3qsai0nz8maad51349v19yqf16rj2yv-alsa-lib-1.1.8' from 'https://cache.nixos.org'...
copying path '/nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/dz580wi3mhj0a217vlyj1hdih3ldiz14-dbus-glib-0.110' from 'https://cache.nixos.org'...
copying path '/nix/store/rg93lv1yxxqw3v9hg1hkpfqr2nab5p7k-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/x8ly9dpv355jndl8wn3br2l50n3kz1s2-libIDL-0.8.14' from 'https://cache.nixos.org'...
copying path '/nix/store/rryr4ihjnnwxr6qlj5x9p1mwyg82d34l-libc++abi-5.0.2' from 'https://cache.nixos.org'...
copying path '/nix/store/7x5sfjddnq51py9h7hdj3369dwnlnfkh-ORBit2-2.14.19' from 'https://cache.nixos.org'...                                                                                                                                                                                                                                        
copying path '/nix/store/s2f809sapq9dcbbgl1gv2yfra119s740-libc++-5.0.2' from 'https://cache.nixos.org'...                                                                                                                                                                                                                                          
copying path '/nix/store/bqyn2l33zzm8pgcd973g46qz534swzm3-liberation-fonts-2.00.4' from 'https://cache.nixos.org'...
copying path '/nix/store/wnxw0bi0jcr729akrh8sqpizdvhdnh9g-ncurses-6.1-20190112-abi5-compat' from 'https://cache.nixos.org'...
copying path '/nix/store/0xy2280dk772d64kvba3ibvcapqn5vl5-platform-tools-28.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/2p5331cg724c69hcfj203ji8sj9bqv2s-polkit-0.115' from 'https://cache.nixos.org'...
copying path '/nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/jsycgv6la5nhsyb2034l7xfc5sa0ajqx-gconf-3.2.6' from 'https://cache.nixos.org'...   
copying path '/nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/vzmvbd94qm3kjv8m7ljbsyy8n3nygmy6-gnome-vfs-2.24.4' from 'https://cache.nixos.org'...  
copying path '/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static' from 'https://gerschtli.cachix.org'...
copying path '/nix/store/nqf6dmfdb7b62bdapysnf2dc4cjp7al1-openjdk-8u212-ga-jre' from 'https://cache.nixos.org'...
building '/nix/store/rp8yadwvsjxpan0lc11h5pc7kxsh1dh9-home.nix.default.drv'...                   
copying path '/nix/store/2iqzwckaprdrr69wy9kap23i94q0bi3w-openjdk-8u212-ga' from 'https://cache.nixos.org'...
building '/nix/store/23bwyp0gb2r21abp93nzhznllycg2vs5-login-inner.drv'...                                   
copying path '/nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045' from 'https://gerschtli.cachix.org'...
building '/nix/store/kby963p6gbm56w43zj4i7yxfwfznzhmb-login-inner.drv'...                                
copying path '/nix/store/hx9mk75m6xb4vc1mczd48pdcmp5if95k-bionic-prebuilt' from 'https://gerschtli.cachix.org'...                                                                                                                                                                                                                                  
copying path '/nix/store/qfg481dl6wal1ik4z7xqlgjnhjkxrkr8-bionic-prebuilt' from 'https://gerschtli.cachix.org'...                                                                                                                                                                                                                                  
copying path '/nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android' from 'https://gerschtli.cachix.org'...                                                                                                                                                                                                         
copying path '/nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android' from 'https://gerschtli.cachix.org'...
building '/nix/store/cz538illxw8cr4a2khvq8gnv2yab9z0w-login.drv'...                                                                                                                                                                                                                                                                                
building '/nix/store/7pyzz8b723q3m78606c9xap5c614gihr-nix-on-droid-install.drv'...                      
building '/nix/store/fflhixzb56iyi94rv0jcaqchsqq6ajbs-nix-on-droid-install.drv'...                                                                           
building '/nix/store/5i6qpb3ypr10wlg6kvi56w5syifkwcwb-nix.conf.drv'...                                       
building '/nix/store/wgch6p4mf35an2g1djnqa8gz9253dlsm-resolv.conf.drv'...                                            
/nix/store/5apw5dbp9bicw4f63xcpxmijc7dk2w9c-bootstrap                                                            
/nix/store/j2j4jy8abzqilay1jcjqjn37cgnapszl-bootstrap-zip                     
/nix/store/n0cwr7553h1qscs7fg3yi5ak0s0kv911-home.nix.default                                                 
/nix/store/13f9dn1dcz6v3nr0rkbg5mi7m0wr23cx-login                        
/nix/store/nkm4zafhqgggvmdk0fvn2y5vx1p508rd-login-inner                                                                  
/nix/store/bj6wbzdf6f4aww23l1dbxalsdsgjqm61-nix.conf                     
/nix/store/01xidc3kn1q2dgr1hs61wbjp8lkpjg45-nix-on-droid-install                                                 
/nix/store/68c85jg2gf0kkwlxhvd6rd5rxph9d9z6-resolv.conf                                                          
/nix/store/3izx77j932vi5g03725rv1qmcdjmsvjn-nixDirectory                                                                                  
/nix/store/4mdhz0bwrak1h29qvli9f8gjz89zbgxp-nix-installer-sha256                                                                             
/nix/store/d7s5dnx687kyigk6qdc65wdwfy0idlff-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android
/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static                   
/nix/store/5zpgwq3xj4rnsx6n6j8himf34f5japl1-talloc-2.1.14-aarch64-unknown-linux-android
/nix/store/0f2lvyfq7hmp63rwbwgxy77wal3xjq69-bootstrap                 
/nix/store/82hax6bnz4gazxapiyb7lz2msv36aw46-bootstrap-zip                
/nix/store/n0cwr7553h1qscs7fg3yi5ak0s0kv911-home.nix.default
/nix/store/13f9dn1dcz6v3nr0rkbg5mi7m0wr23cx-login        
/nix/store/ym08s9gw48kx3nrsl6rxjvhrj0jsfvvm-login-inner     
/nix/store/bj6wbzdf6f4aww23l1dbxalsdsgjqm61-nix.conf
/nix/store/lvi50fymscyj4m1a8h12ia9j6m8xxbak-nix-on-droid-install
/nix/store/68c85jg2gf0kkwlxhvd6rd5rxph9d9z6-resolv.conf
/nix/store/zyxl5gmrhy9gql4cw7hpv05hz947aal0-nixDirectory        
/nix/store/y6viscr0ypik5ihhk3sksmx99z9zwdmh-nix-installer-sha256
/nix/store/5rk098j6ib0jv49438i44jvl9lf68cyp-proot-termux-unstable-2019-09-05-i686-unknown-linux-android
/nix/store/mhqv5dzzzpzq8p49hhxnwi6q1rs5jjhz-qemu-aarch64-static 
/nix/store/65x9h7hnby3g2lrv7apj1wz2f59r5399-talloc-2.1.14-i686-unknown-linux-android 
```

Can you reproduce this, @t184256 ?

### CODE REVIEW

**Comment by Gerschtli** [comment]
Yes, I was right. `writeText` and `writeScript` are always built locally, see https://github.com/NixOS/nixpkgs/blob/05393ca2170bf5db031372bdf2a714b006f9b2ca/pkgs/build-support/trivial-builders.nix#L64

**Comment by Gerschtli** [comment]
With my last commit https://github.com/t184256/nix-on-droid-bootstrap/pull/15/commits/13e083806e8333f57b90dfc15a6be062f0f56b82 (removed it already via force push) travis succeded to build: https://travis-ci.org/t184256/nix-on-droid-bootstrap/builds/588028928.

Therefore I hope it fixes your local build issues. The only thing left to do is setting `CACHIX_SIGNING_KEY` via the travis-ci web gui (https://travis-ci.org/t184256/nix-on-droid-bootstrap/builds/588028928#L557).

**Comment by t184256** [comment]
I think that the real problem is that the hashes of the results that I build differ from what your builds and Travis builds. That's strange.

**Comment by Gerschtli** [comment]
Oh, that is very strange indeed.. Do you have some crazy things in `~/.config/nixpkgs/config.nix` or `~/.config/nixpkgs/overlays*`?

**Comment by t184256** [comment]
Oh yeah, I do have a metric ton of overlays.

I think I've found a culprit, and it was an innocent looking `environment.noXlibs = true;` in my system config. Rebuilding without it to check that it's the only one.

Do you what to do to make the builds ignore my system config altogether? I thought that I've sidestepped that altogether with nixpkgs pinning, but, apparently, that was not enough.

**Comment by Gerschtli** [comment]
Try it again. I think it should work now.

**Comment by t184256** [comment]
Yeah, seems like it helped!

**Comment by t184256** [comment]
How long did it take for you to push `/nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045` to cachix? I've tried pushing from two different countries, waiting for 1.5+ hours, and it still looks like it just hangs =/

**Comment by Gerschtli** [comment]
Ah it's a very big store path indeed..
```
$> nix path-info /nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045 -s           
/nix/store/qdi2jfijw9p230bz113hmvhcz4pai69x-ndk-bundle-18.1.5063045	 1991535552
```

1991535552 B = ~ 1.8 GB

Depending on your upload speed it could take a while :D

File: src/pkgs/nix-directory.nix [hunk 1]
```diff
@@ -34,9 +21,11 @@ in
 buildPkgs.stdenv.mkDerivation {
   name = "nixDirectory";
 
+  # FIXME: find a source where sha256 never changes, nixInstallerSha256
```
Comment (t184256) [inline src/pkgs/nix-directory.nix]: Could you elaborate on this one? Why do you think it changes? Why not just add constant hashes per arch?

**Comment by t184256** [comment]
The build passed!

To be honest, I consider uncovering the hash mismatch and disabling the impurity stemming from my build machine configuration to be even cooler than introducing a CI. Great job on both of these!

File: src/pkgs/nix-directory.nix (hunk 1 above)
Comment (Gerschtli) [reply src/pkgs/nix-directory.nix]: When I think about it, we could probably hardcode the sha256 per arch. I was just thinking that this were rolling releases like the nix channel tar balls, but this one will probably never get updated.. My fault! I can fix it in the next days or you can do it, it's up to you. Thanks for pointing this out :)

**Comment by Gerschtli** [comment]
Haha, yeah, back in days when I had troubles with my local overlays when building packages of my local nixpkgs, it took quite much time to figure out these options :D

Next step (what isn't as easy as I thought): Building the initial home-manager generation and put it in the zip ball.

I moved home dir to `PREFIX/usr/home` without problems, but `nix-channel --update` don't want to run successfully in `proot`..

**Comment by t184256** [comment]
> Next step (what isn't as easy as I thought): Building the initial home-manager generation and put it in the zip ball.

I'm not a fan of this direction, and here's why. I intend to ship is Nix for Android and I want advanced users to have exactly this. I see value in the initial zipball being just a repack of the official image.

Now another thing is that my current methods are sorely lacking. Those static bundled files are necessary clutches, and, is you have noticed, they don't age well. I like the idea of overwriting them with a home-manager module and managing them with h-m since that point, but I also want a clean base image and an opt-in for those who are fine with bare nix. Forcing a 0.5 GB zipball intricately and complicatedly tied to h-m out of the box... Not so sure about that.

Perhaps I should learn a thing or two about the internals of this nix-darwin thing. It seems to solve the same idea of shipping Nix on an uncooperative host system, and it has probably solved that dilemma long ago.

**Comment by Gerschtli** [comment]
I see your point there. But if we don't force home-manager (and the future possibility to manage the /bin/sh and /usr/bin/env symlinks and the custom scripts in /bin), there has to be an easy way to update these.

Maybe we can just provide a script that is run as post-build-hook (new in nix 2.3) that will just update the files and symlinks. As I'm writing this, this seems to be the best way to handle it. I will do some testing to see if this works like indented.

That way there would be no need for a nix-on-droid home-manager module nor any home-manager installation.

**Comment by t184256** [comment]
Managing them with some separate and small parametrised derivation that gets installed into the profile and may or may not be managed with h-m sounds the best.

I wonder, why would we need post-build-hooks for updating these files? Why won't symlinking them somewhere into the profile directory be enough?

**Comment by Gerschtli** [comment]
A custom derivation providing all the necessary scripts and packages like `coreutils` and `bashInteractive` would also be a good solution. Linking `/bin/sh` to `~/.nix-profile/bin/sh` and the same for `/usr/bin/env` and moving the scripts and binaries to `~/.nix-profile` as well. Maybe the other files in `/etc` too. Did you mean something like this? 

How do we want to achieve this?
We could provide an overlay for this derivation. With that you would have maximum control over the derivation. For that we have to run `nix-channel --update` during build phase of the zip ball. After that install the overlay and then this new derivation. Do you have a better approach to this?

The idea with `post-build-hooks` won't work as they get called after *every* build.

**Comment by t184256** [comment]
Short answer: yes, but with on-device execution and symlinks initially pointing outside the profile.

Long answer:

I am not that well-versed with nix. My idea goes like this:

Only put whatever is needed to run nix in the initial zipball. Symlink to the required 'stub files' from `/d/d/c.t.n/f/usr/bin` -or- symlink the entire `/d/d/c.t.n/f/usr/bin` to the directory with the stub files.

On the 'first boot', detect that fact and build a derivation that provides more meaningful content of them with the initial bundled nix. Install it to the user profile and update the symlinks to point to profile contents.

Then give user a choice to either proceed from that on their own with caution or install home-manager, which will ship with a config pre-depending on this derivation and demonstrating how to parametrize it.

Making this derivation live in an overlay sounds like a very good idea, with pros including introducing an actual upgrade path that doesn't require wiping/reinstallation, and cons including introducing an way to accidentally break installs remotely =). Would be nice to provide a revert mechanism from within the recovery shell, mmm.

Not sure how to feel about it making it pull in coreutils and bashInteractive into `$PATH`. 'Why not', I guess.

So, pretty much what you have suggested, but without baking this into the initial zipball and with initial links pointing somewhere else. Why go the way of solving on-host in-proot execution and bloating the zipball size by bundling stale derivations if the same could be achieved in reasonable time on 'first boot'?

**Comment by Gerschtli** [comment]
This seems to be an easier way to implement this.

Let's see when I got time for this :)

**Comment by Gerschtli** [comment]
I found a better solution than an overlay: a nix-on-droid nix-channel. The user has full control over the provided version with rollback and manual updates!

**Comment by t184256** [comment]
I don't fully understand your thought. As far as I understand, a channel is just a delivery method, while a deliverable could be just derivations, overlays, entire system management mechanisms like home-manager... My thoughts are either just a derivation to provide the required files and something closer a home-manager module, if it is even possible to deliver its modules in separate channels and not push them upstream.

**Comment by Gerschtli** [comment]
If you implement it as derivation/overlay/whatever, you still need a way to provide updates and rollbacks. This can be achieved by fetching nix-on-droid as nix channel. Of course there could be still a home-manager module.

**Comment by t184256** [comment]
Yes, shipping it in a separate channel makes sense anyway, so that you can update it separately and pin it. Can't and won't argue with that. 

Somewhat independently of pinning there's a rollback mechanism in home-manager that can be used to revert to past instantiations. That was what I had in mind when I talked about reverts, and it's basically already there, we just need to a) manage more stuff through nix on-device b) build a relatively unbreakable way to invoke it through a recovery shell, which does not seem so trivial to me.

**Pull request merged**